File: hackerrank/twins/twins.py
from math import floor

def count_twins(n, m):

	# segmented sieve ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
	# use simple sieve to compute primes upto sqrt(m)
	isprime = [True] * (m+1)
	isprime[0] = isprime[1] = False

	d = 2
	while d*d <= m:
		if isprime[d]:
			i = 2
			while i*d <= m:
				isprime[i*d] = False
				i += 1
		d += 1

	lo = floor(m**0.5) + 1
	hi = 2 * (floor(m**0.5) + 1)

	while lo < m:
		if hi >= n:
			hi = n



	# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

	count = 0

	for i in range(n, m-1):
		if isprime[i] and isprime[i+2]:
			count += 1

	return count

def main():
	n, m = map(int, input().split())
	print(count_twins(n, m))

main()

File: hackerrank/twins/twins.cpp
#include <iostream>
#include <vector>
#include <cstring>
#include <cmath>
using namespace std;

typedef unsigned long long ull;

void count_twins(ull n, ull m){
	vector<bool> isprime(true, m+1);
	isprime[0] = false;
	isprime[1] = false;

	for (ull i = 2; i*i <= m; ++i){
		if (isprime[i]){
			ull j = 2*i;
			while (j <= m){
				isprime[j] = false;
				j += i;
			}
		}
	}
	ull count = 0;
	for (ull i = n; i <= m-2; ++i){
		if (isprime[i] && isprime[i+2]){
			count++;
		}
	}

	cout << count << endl;

}

void simple_sieve(ull limit, vector<ull>& primes){
	bool mark[limit + 1];
	memset(mark, true, sizeof(mark));
	for (ull p = 2; p*p < limit; ++p){
		if (mark[p]){
			for (ull i = p*2; i < limit; i += p){
				mark[i] = false;
			}
		}
	}

	for (ull p = 2; p < limit; ++p){
		if (mark[p])	primes.push_back(p);
	}

}

void segmented_sieve(ull n){
	ull limit = floor(sqrt(n)) + 1;
	vector<ull> primes;
	simple_sieve(limit, primes);
	ull lo = limit;
	ull hi = 2 * limit;

	while (lo < n){
		if (hi >= n)	hi = n;
		bool mark[limit + 1];
		memset(mark, true, sizeof(mark));

		for (ull i = 0; i < primes.size(); ++i){
			ull loLim = floor(lo/primes[i]) * primes[i];
			if (loLim < lo){
				loLim += primes[i];
			}

			for (ull j = loLim; j < hi; j += primes[i]){
				mark[j - lo] = false;
			}
		}

		/*
		for (ull i = lo; i < hi; ++i){
			if (mark[i - lo])
				cout << i << " ";
		}
		*/

		lo += limit;
		hi += limit;

	}
	cout << endl;
}

int main(){
	ull n, m;
	cin >> n >> m;
	count_twins(n, m);

//	vector<ull> primes;
//	simple_sieve(1e9, primes);

	//segmented_sieve(1e9);

	return 0;
}

